### PR TITLE
Add junit tests

### PR DESCRIPTION
**Pull request by sdonkov**
#14

### CODE REVIEW

File: src/main/java/PrefixTree.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import java.util.ArrayList;
```
Comment (dmitankin) [inline src/main/java/PrefixTree.java]: This file has no place here. As I mentioned previously, please self-review before placing a PR. You are wasting the reviewer's time.

**Review by dmitankin — COMMENTED** [review]
I still don't understand why do you insist on putting "Topic" in the description of the PR.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                words.add(currentWord);
+                prefixTree.add(currentWord);
+            }
+        }
+    }
+
+    @Test
+    void addingWordsToTree() {
+        words.forEach(s -> assertTrue(prefixTree.contains(s),
+                "Tree doesn't contain this word - " + s));
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: Here and below, this would lead to creating a new string for each word being checked. The string will be used only upon failure. Creating a new string is expensive relative to the check, so it'd slow the test significantly when the number of checked words grows as planned.
I guess it'd be best to simply pass the checked word instead of a coherent message. This way it'd be apparent which word is causing a failure without a performance penalty.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            while (sc.hasNextLine()) {
+                CharSequence currentWord = sc.nextLine();
+                words.add(currentWord);
+                prefixTree.add(currentWord);
+            }
+        }
+    }
+
+    @Test
+    void addingWordsToTree() {
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: The "ToTree" in the name is redundant. This class contains tests only for the PrefixTree, after all.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        words.forEach(s -> assertTrue(prefixTree.contains(s),
+                "Tree doesn't contain this word - " + s));
+    }
+
+    @Test
+    void negativeTestPrefixTree() throws FileNotFoundException {
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: This name is not descriptive enough as "negative" may mean a lot of different things in this context.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+
+public class PrefixTreeTest {
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java:16]: It makes sense to test edge cases for all public methods of the class, e.g. `null` and empty string.
Also, adding an already existing word.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    void testAddEmptyString() {
+        assertTrue(prefixTree.add(""));
+    }
+
+    @Test
+    void testAddAlreadyAddedWord() {
+        assertFalse(prefixTree.add("table"));
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: This test should instantiate its own trie; otherwise, it depends on the contents of the file which is unnecessary and may cause unprovoked failures when the input is changed. Better test in isolation as much as possible. Applies to the newly added tests above too.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            while (sc.hasNextLine()) {
+                String currentWord = sc.nextLine();
+                assertFalse(prefixTree.contains(currentWord), currentWord);
+            }
+        }
+    }
+
+    @Test
+    void testContainsThrowsAppropriateException() {
+        assertThrows(NullPointerException.class, () -> prefixTree.contains(null));
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: It seems to me that a better response is to return `false` instead of throwing; otherwise, you'd force consumers to always check their input which will only encumber their code with no added value.
The best course of action would be to leave the test like this for now and make this change in a separate PR.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        try (Scanner sc = new Scanner(fileNegative)) {
+            while (sc.hasNextLine()) {
+                String currentWord = sc.nextLine();
+                assertFalse(prefixTree.contains(currentWord), currentWord);
+            }
+        }
+    }
+
+    @Test
+    void testContainsThrowsAppropriateException() {
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: `testContainsNull` is more appropriate really. The names of most of the methods here do not contain info about the expectation. Applies to other exception case too.

**Review by dmitankin — COMMENTED** [review]
Please add an empty test for the contains method too

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    @Test
+    void testAddNull() {
+        PrefixTree underTest = new PrefixTree();
+        assertThrows(NullPointerException.class, () -> underTest.add(null));
+    }
+
+    @Test
+    void testAddEmptyString() {
+        PrefixTree underTest = new PrefixTree();
+        assertTrue(underTest.add(""));
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java:56]: It'd be nice to assert here that `contains` returns false even after "adding" it.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+
+public class PrefixTreeTest {
+
+    private static List<CharSequence> words;
+    private static PrefixTree prefixTree;
+
+    @BeforeAll
+    static void setUp() throws FileNotFoundException {
+        File file = new File("src/test/resources/tests.txt");
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: Here and below, instead of hardcoding this path, you should be able to use `getResourceAsStream("tests.txt")` now that the file is under resources. See [this thread](https://stackoverflow.com/questions/15749192/how-do-i-load-a-file-from-resource-folder).

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    private static List<CharSequence> words;
+    private static PrefixTree prefixTree;
+
+    @BeforeAll
+    static void setUp() throws IOException {
+        words = new ArrayList<>();
+        prefixTree = new PrefixTree();
+        try (InputStream is = PrefixTreeTest.class.getResourceAsStream("/tests.txt")) {
+            if (is != null) {
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: Here and below, there is usually no reason to write test code defensively. On the contrary, this may lead to passing tests when in reality things have messed up. If `is` is `null`, it means we can't test, i.e. we need the crash.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                Scanner sc = new Scanner(is);
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: An alternative to the `Scanner` in this case is the [readAllLines](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#readAllLines-java.nio.file.Path-) method, which would be nicer as code-readers won't have to figure out what do you need a `Scanner` for (it has other uses). Unlike the `Scanner`, it also ties nicely with lambdas, which you can utilize here too, instead of looping.

File: pom.xml [hunk 13]
```diff
@@ -12,5 +12,40 @@
...
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.9.1</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
```
Comment (dmitankin) [inline pom.xml]: I haven't looked into this really, just wondering. Any idea why do we need the compiler plugin now? We were compiling before having tests too.

File: src/test/java/com/github/sdonkov/PrefixTreeTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+    private static List<CharSequence> words;
+    private static PrefixTree prefixTree;
+
+    @BeforeAll
+    static void setUp() throws IOException, URISyntaxException {
+        words = new ArrayList<>();
+        prefixTree = new PrefixTree();
+        List<String> lines = Files.readAllLines(Paths.get(PrefixTreeTest.class.getResource("/tests.txt").toURI()));
+        words.addAll(lines);
```
Comment (dmitankin) [inline src/test/java/com/github/sdonkov/PrefixTreeTest.java]: Is this because `words` is not `List<String>`?
You can simply make it so, nothing will break as far as I can tell and you'll avoid declaring `lines`. 
If you are curious on how `addAll` here works see Effective Java's "Item 31: Use bounded wildcards to increase API flexibility" - it's among the challenging parts of the book.

File: .github/workflows/test.yaml [hunk 15]
```diff
@@ -0,0 +1,89 @@
+# This workflow will build a Java project with Maven, and cache/restore any dependencies to improve the workflow execution time
```
Comment (dmitankin) [inline .github/workflows/test.yaml]: This file deserves its own PR, I think.

**Review by dmitankin — APPROVED** [review]
There's one comment on `testAddEmptyString` still standing.
Feel free to ignore it and check this in (minus the new test.yaml thing).
